pkg/endpoint/local_ip.go

```diff
@@ -44,21 +44,41 @@ func getLocalIPFromRoutes(family k8snet.IPFamily) (string, error) {
 	}
 
 	for i := range routes {
-		if routes[i].Gw != nil {
-			return routes[i].Src.String(), nil
+		if routes[i].Gw != nil && routes[i].Src != nil {
+			// OVNK configures special internal masquerade IPs,
+			// these IPs are used internally by OVNK infra dataplane, and they should
+			// not be selected as private IPs.
+			ip := routes[i].Src
+			if family == k8snet.IPv6 && !isValidGlobalIPv6(ip) {
+				continue
+			}
+
+			return ip.String(), nil
 		}
 	}
 
 	return "", errors.Errorf("couldn't find route for family %v", family)
 }
 
 func GetLocalIPForDestination(dst string, family k8snet.IPFamily) string {
-	conn, err := Dial("udp"+string(family), "["+dst+"]:53")
+	conn, err := Dial("udp"+string(family), net.JoinHostPort(dst, "53"))
 	if err == nil {
 		defer conn.Close()
 		localAddr := conn.LocalAddr().(*net.UDPAddr)
 
-		return localAddr.IP.String()
+		if family == k8snet.IPv4 || isValidGlobalIPv6(localAddr.IP) {
+			return localAddr.IP.String()
+		}
+
+		logger.Warningf("IP %q returned from Dial isn't usable - trying to find another IP from the same interface", localAddr.IP)
+
+		// Try to find a valid global-scope IP from the same interface
+		ifaceIP, ifaceName, err := getValidGlobalIPv6FromSameInterface(localAddr.IP)
+		if err == nil {
+			return ifaceIP
+		} else {
+			logger.Warningf("Failed to retrieve valid IPv6 address from interface %q for %q", ifaceName, localAddr.IP)
```

Comment:
We should log the error message here:

```suggestion
			logger.Warningf("Failed to retrieve valid IPv6 address for %q: %v", ifaceName, localAddr.IP, err)
```